src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala

```diff
@@ -212,15 +344,131 @@ class Elementwise(val param:LayerParameter, val id:Int, val net:CaffeNetwork) ex
   
 }
 
+class Concat(val param:LayerParameter, val id:Int, val net:CaffeNetwork) extends CaffeLayer {
+  override def sourceFileName = null
+  override def init(dmlScript: StringBuilder): Unit = {}
+  var _childLayers:List[CaffeLayer] = null
+  
+  // Utility function to create string of format:
+  // fn(fn(fn(_childLayers(0).out, _childLayers(1).out), _childLayers(2).out), ...)
+  // This is useful because we do not support multi-input cbind and rbind in DML.
+  def _getMultiFn(fn:String):String = {
+    if(_childLayers == null) _childLayers = net.getBottomLayers(param.getName).map(l => net.getCaffeLayer(l)).toList
+    var tmp = fn + "(" + _childLayers(0).out + ", " + _childLayers(1).out + ")"
+    for(i <- 2 until _childLayers.size) {
+      tmp = fn + "(" + tmp + ", " +  _childLayers(i).out + ")"
+    }
+    tmp
+  }
+  
+  /*
+   * Computes the forward pass for a concatenation layer.
+   *
+   * Inputs:
+   *  - n_i * c_i * h * w for each input blob i from 1 to K.
+   *
+   * Outputs:
+   *  - out: Outputs, of shape 
+   *    - if axis = 0: (n_1 + n_2 + ... + n_K) * c_1 * h * w, and all input c_i should be the same.
+   *    - if axis = 1: n_1 * (c_1 + c_2 + ... + c_K) * h * w, and all input n_i should be the same.
+   */
+  override def forward(dmlScript: StringBuilder, isPrediction: Boolean): Unit = {
```

Comment:
We should just add a new `nn/layers/concat.dml` layer, rather than embed it in this Scala code.  That way, we can test it for correctness using the same grad checks the rest of the `nn` library uses, and can use it in other contexts.